### ISSUE TITLE
Have the buildpack run `poetry lock` if no lockfile found?

### ISSUE THREAD

**Issue by kavdev** [moneymeets/python-poetry-buildpack#32]
What are your thoughts on having this buildpack run `poetry lock` if it doesn't find an existing `poetry.lock` file?

I have dependabot running on my GitHub repo. Each time I merge a dependabot PR, it forces a cascade rebase of all other dependency PRs because of lockfile hash conflicts. It'd save me a ton of time (and CI/CD credits) if I could not commit the lockfile and just have it be generated at runtime.

**Comment by stohrendorf** [comment]
That would bring in the disadvantage of non-reproducible builds, which in turns makes it harder to identify causes of bugs, which may be caused by bugged dependencies or just plain errors in your CI pipelines. If you have, for example, the suspicion that your CI pipeline is at fault, you'd loose the guarantee that repeatedly triggering a pipeline for a commit uses the very same dependencies, because between two runs a dependency might have been updated, changing the behavior of the pipeline without further notice. The worst thing I can imagine about this is that a dependency changes its behavior, which in turn starts to disguise a bug that's actually on your side.

I understand that you want to get updates "for free", but IMHO that only runs fine if there's not a single bug in all your code paths, including the code paths triggered in 3rd party dependencies. If there's a bug introduced, or you start to use a code path that you never used before but contains a bug, you don't have the guarantee that you can analyze this bug. Even worse than that, you have no guarantee at all that all your developers share the same environment, i.e. what works for one developer may not work for others.

That's a lot of cons I see here, but I'm curious what pros there are (because I currently can't think of any).

**Comment by kavdev** [comment]
All of our dependencies (and key sub-dependencies) are pinned, and we're willing to take a risk on non-key sub-dependencies. This is less a discussion on the pros and cons of a committed lockfile -- that's the standard practice and recommended by poetry.

We just happen to find ourselves in a situation where dependabot PRs are no longer very useful and are costing us a ton of CI/CD minutes. The solution for us would be to ditch the lockfile and rely on pinned dependencies in the pyproject.toml (similar to as we had before with straight requirements.txt files).

I didn't say it was a _great_ idea 🙂; I'm just wondering if you'd be up for supporting it as a fallback (or with an env flag).

**Comment by stohrendorf** [comment]
Ah, sorry, I misunderstood the intended scope. I guess we'll talk about this in our team, but I can't give an estimation when this will be done.

**Comment by zyv** [comment]
Hey @kavdev,

TLDR; I have tried to implement your feature request, but came to the conclusion that there is no good way to support it in this buildpack without investing an inordinate amount of work and dramatically increasing the complexity without compromising on safety and usability for "normal" workflows:

https://github.com/moneymeets/python-poetry-buildpack/compare/experiment/no-poetry-lock

## detect

First, we would need to change the `detect` script to check for `pyproject.toml` instead of `poetry.lock`. However, `pyproject.toml` is a generic project format and doesn't necessarily imply that the project is managed by Poetry. Only `poetry.lock` is a proprietary Poetry file, which unambiguously indicates that the project is being managed by Poetry.

Unfortunately, the buildpack API doesn't allow accessing environment variables from `detect`. So we cannot just require **either** `poetry.lock`, **or** `DISABLE_POETRY_LOCK_CHECK=1`.  Either we allow breakage in `detect`, or we need to look for `poetry.lock` first and then if it's not there, look for `pyproject.toml` instead to parse it in Bash to make sure it has valid Poetry sections - be my guest, tons of fun... or else install Poetry and run `poetry check` - not sure it's even possible, but definitely huge slowdown and waste of bandwidth & resources.

And, by the way, now that `detect` has become rather complicated, one needs to write some good tests for it. Alright, we can just check for `pyproject.toml` for now and try to catch the breakage in `compile`...

## compile

Now, we have to introduce a check for `poetry.lock`, because we normally don't want to have people forget to commit their locks. Which means that we have to introduce an opt-out, e.g. `DISABLE_POETRY_LOCK_CHECK`. So far, so good. Of course, this stuff has to be tested, so let's add some basic tests.

Now comes the annoying part: we normally generate `runtime.txt` from `poetry.lock`. So you would also have to opt out via `DISABLE_POETRY_CREATE_RUNTIME_FILE`, or set Python version using `PYTHON_RUNTIME_VERSION`. 

So we need to add another check - if `DISABLE_POETRY_LOCK_CHECK` is set, then either `DISABLE_POETRY_CREATE_RUNTIME_FILE` or `PYTHON_RUNTIME_VERSION` has to be set as well. Of course, all these code paths have to be carefully tested as well.

At last, some documentation would be in order...

---

At this point I've given up: as you can see, I've really went out of my way trying, but it all feels so completely wrong, annoying and overcomplicated... If you are willing to give it another go, we might consider merging it if the tests are really *perfect*, but I hope you now understand what I meant in my first sentence.

Now, to be honest, it seems to me that your workflow is somehow broken, which leads to Dependabot not working well for you (or maybe Dependabot is broken as well ;-) ) and you want to fix this brokenness by adding more breakage to the deployment process.

First off, pinning major dependencies in `pyproject.toml` is generally a bad idea. It is meant to be used for pinning semantic versions (e.g. `python = ">= 3.8"` for walrus), and not specific versions. Specific versions are meant to be updated by executing `poetry update package_name` for a specific package or `poetry update` for the complete dependency tree. These commands are not meant to be used in a normal workflow without thinking & testing the results...

I suspect, that this is why you are having problems with Dependabot in the first place. It should be able to resolve lock conflicts, but probably it has problems rebasing `pyproject.toml`.

Solution: fix your workflow, educate your developers, add merge checks for specific commit messages upon lock updates.

Second, as I say, Dependabot should be normally able to rebase pull requests that it creates when you update lock files in your master branch. If it has problems rebasing, one should see if they can fix it. Make sure that you are using the right Dependabot, by the way. Many still use the old one, and wonder how come it doesn't work well.

Finally, if you don't want or cannot change anything, you could also ask Dependabot to create only one PR and wait until it's merged for the next one. That should actually work fine most of the time, unless there are some urgent security updates.

Hope that helps!

P.S. You can of course also fork my experiment branch and use this at your own peril if you like.

**Comment by ryanhiebert** [comment]
Folks here may be interested in this issue: https://github.com/dependabot/dependabot-core/issues/3212

Basically, I'm finding that dependabot is really noisy, and because it updates caret requirements unnecessarily, it causes a lot of extra builds for my application as the conflicts have to be resolved.

I am running an experiment right now to change the caret requirement to a ~ requirement, to see if dependabot knows how to work with that better. I expect I'll have some kind of answer on that tomorrow, as one of our dependencies gets updated virtually every night (which drives me nuts, boto3).

**Comment by zyv** [comment]
@ryanhiebert thanks for the notice - I think now I can understand the OP's problem better - and the reason we are quite happy is that we don't use caret requirements and instead go full `*`. Not sure if this can help you as well, but good thing that you have reported it anyways - this is the only way to get it fixed. Maybe if you provide dependabot people with a few examples, they'll be able to figure it out - if that's happening to the repositories that are public, of course.

**Comment by ryanhiebert** [comment]
Thanks, a `*` requirement might be even better. My experiment with `~` did work, but I haven't yet determined whether `~` requirements work as PEP 440 specifies for a [compatible release](https://www.python.org/dev/peps/pep-0440/#compatible-release), or as Poetry's docs specify for [tilde requirements](https://python-poetry.org/docs/versions/#tilde-requirements). It surprised me to find that they are documented with different semantics. Even more I don't know what Dependabot does with them.

**Comment by kavdev** [comment]
@zyv thanks for taking the time to look into this. I really appreciate the writeup.

The issue isn't that dependabot can't rebase -- it's that the lockfile hash is different for each change, so merging one PR causes each other PR to rebase and run through CI/CD again.

I'll go ahead and close this. There's a possibility this might get resolved on the [poetry side of things](https://github.com/python-poetry/poetry/issues/496) with a change in the way the main file hash is handled, but we'll see.

**Comment by zyv** [comment]
> The issue isn't that dependabot can't rebase -- it's that the lockfile hash is different for each change, so merging one PR causes each other PR to rebase and run through CI/CD again.

@kavdev thanks for your patience and following up on that - I think I have now finally completely understood your problem - better late than never :-)

Unfortunately, at this point I can only suggest do use my experimental branch, if you really need to save CI time until this is fixed in Poetry - we are batching our Dependabot updates, so this doesn't really affect us, and supporting your workflow turned out to be much more complicated than we could afford...

I have voted for the Poetry request though, and invite everybody else to do the same ;-)
